### ISSUE TITLE
Silent renew process over PKCE seems to use renewTimeBeforeTokenExpiresInSeconds not the way you would expect - possibly a BUG

### ISSUE THREAD

**Issue by silviu-sterian** [damienbod/angular-auth-oidc-client#1515]
**What Version of the library are you using?**
Using v 14.1.3

**Question**
How can I trigger the renewal of the token a certain number of seconds before it expires?
The recommended way is setting renewTimeBeforeTokenExpiresInSeconds = xx but this seems to just give an offset to the expiration time.  I wish it would trigger the silent renew xx seconds before token comes back as expired, not expiring the token sooner

In short, I looked at the code in TokenValidationService -> validateAccessTokenNotExpired(..)  and confirmed the logic is offsetting the expiration time but all it does is expire it sooner. So it will come back as expired during this check early (by the number of seconds). 
This does not help closing the small window where token shows as expired and if a user clicks on anything and we're trying to get the token for an api call it will trigger the authentication process - round-trip to the identity provider.

What I was hoping I can get is an early trigger of the silent renewal process - by a number of seconds, while the token is still valid for the number of seconds. Can you help me with an workaround or let me know if you;re considering adding this feature?

**At this point I am pretty sure it is a bug. Ideally we should either change the function calling validateAccessTokenNotExpired(..)  to not pass in the offset and just trigger the process early or give a user of the library more control to trigger the process based on a custom trigger.**

The screenshot attached is representative of the issue. The expiration time should not be negative: 

![image](https://user-images.githubusercontent.com/38025318/185714054-ea53fe83-1a61-4669-8cea-434b9c821a15.png)

**Comment by silviu-sterian** [comment]
![image](https://user-images.githubusercontent.com/38025318/185658482-b0d61737-b76e-4372-9e32-c2b3a08d2ba6.png)

**Comment by andreaslarssen** [comment]
Anyone of the maintainers have a comment on this? I'm suspecting a similar experience that I haven't gotten around to debug yet.

**Comment by dabbid** [comment]
Same for me. Without this option, I cannot get silent renew working, while setting this value to 60s with another library (which is now archived, so my switch to this one) was working well.

**Comment by andreaslarssen** [comment]
@damienbod Can you confirm that this is a bug?

**Comment by damienbod** [comment]
@andreaslarssen  I need to look at this first and reproduce to verify

Greetings Damien

**Comment by silviu-sterian** [comment]
Thank you for looking into this @damienbod
Let me know if you need any info, it should be straight forward to reproduce.

**Comment by silviu-sterian** [comment]
Any update on this?

**Comment by andreaslarssen** [comment]
Seems like a fix could be merged (#1529) ?

**Comment by silviu-sterian** [comment]
> @silviu-sterian Seems like a fix has been merged (#1529). Not sure if it's out, or when it will be available

Thanks @andreaslarssen it looks like it might fix it. I'll pull this fix and test it. What is the release schedule for this?

**Comment by andreaslarssen** [comment]
@silviu-sterian I'm not a maintainer, just waiting for the release like you, but seems like it's out:
https://github.com/damienbod/angular-auth-oidc-client/releases/tag/release_14_1_5

**Comment by silviu-sterian** [comment]
@andreaslarssen  this unfortunately does not fix my issue.  This issue (reported above) has to do with renewTimeBeforeTokenExpiresInSeconds property and it is not solved by this release

**Comment by andreaslarssen** [comment]
> @andreaslarssen  this unfortunately does not fix my issue.  This issue (reported above) has to do with renewTimeBeforeTokenExpiresInSeconds property and it is not solved by this release

Ah, that's too bad. Pretty sure we renew the token before expiration though

**Comment by silviu-sterian** [comment]
@andreaslarssen Thanks again for contributing and keeping this active. I explained how this issue is happening. Even if you might be lucky enough that your library renews the token right when it expires, I don't think currently renewTimeBeforeTokenExpiresInSeconds  is honored, which means you should be able to renew the token - for safety - that number of seconds before it expires. Say you want to make sure you account for latency in your network and set it to 20s. Does the silent renew process start 20s before the token expires?

**Comment by andreaslarssen** [comment]
> @andreaslarssen Thanks again for contributing and keeping this active. I explained how this issue is happening. Even if you might be lucky enough that your library renews the token right when it expires, I don't think currently renewTimeBeforeTokenExpiresInSeconds is honored, which means you should be able to renew the token - for safety - that number of seconds before it expires. Say you want to make sure you account for latency in your network and set it to 20s. Does the silent renew process start 20s before the token expires?

We've had quite few problems with this package lately, so I don't remember all the details, but I thought this indeed fixed the issue of the refresh token actually being refreshed `renewTimeBeforeTokenExpiresInSeconds` seconds before it expired

**Comment by silviu-sterian** [comment]
> > @andreaslarssen Thanks again for contributing and keeping this active. I explained how this issue is happening. Even if you might be lucky enough that your library renews the token right when it expires, I don't think currently renewTimeBeforeTokenExpiresInSeconds is honored, which means you should be able to renew the token - for safety - that number of seconds before it expires. Say you want to make sure you account for latency in your network and set it to 20s. Does the silent renew process start 20s before the token expires?
> 
> We've had quite few problems with this package lately, so I don't remember all the details, but I thought this indeed fixed the issue of the refresh token actually being refreshed `renewTimeBeforeTokenExpiresInSeconds` seconds before it expired

I've upgraded the library and the only change I noticed was that silent renew did not log to console how it used to, although the config settings are identical - LogLevel.Debug

**Comment by andreaslarssen** [comment]
> > > @andreaslarssen Thanks again for contributing and keeping this active. I explained how this issue is happening. Even if you might be lucky enough that your library renews the token right when it expires, I don't think currently renewTimeBeforeTokenExpiresInSeconds is honored, which means you should be able to renew the token - for safety - that number of seconds before it expires. Say you want to make sure you account for latency in your network and set it to 20s. Does the silent renew process start 20s before the token expires?
> > 
> > 
> > We've had quite few problems with this package lately, so I don't remember all the details, but I thought this indeed fixed the issue of the refresh token actually being refreshed `renewTimeBeforeTokenExpiresInSeconds` seconds before it expired
> 
> I've upgraded the library and the only change I noticed was that silent renew did not log to console how it used to, although the config settings are identical - LogLevel.Debug

I'm not currently working on the project where we use this package, but @erichjsonfosse and @ccondrup are. Maybe they can check / shed some light on this issue and see if we face the same

**Comment by damienbod** [comment]
This should be fixed., A lot of these issues is due to us validating the id_token and the id_token does not get refreshed and so expires.

This is fixed in version 14.2.0

Once released you can configure it something like this:

```typescript
import { NgModule } from '@angular/core';
import { AuthModule, LogLevel } from 'angular-auth-oidc-client';

@NgModule({
  imports: [
    AuthModule.forRoot({
      config: {
        authority: '--idp--',
        redirectUrl: window.location.origin,
        postLogoutRedirectUri: window.location.origin,
        clientId: '--client_id--',
        scope: 'openid profile offline_access',
        responseType: 'code',
        silentRenew: true,
        useRefreshToken: true,
        ignoreNonceAfterRefresh: true, // this is required if the id_token is not returned
        allowUnsafeReuseRefreshToken: true, // this is required if the refresh token is not rotated
        enableIdTokenExpiredValidationInRenew: false, // required to refresh the browser if id_token is not updated after the first authentication
        autoUserInfo: false, // if the user endpoint is not supported
        logLevel: LogLevel.Debug,
      },
    }),
  ],
  exports: [AuthModule],
})
export class AuthConfigModule {}
```

If you only need to disable thr id_token check and use only the access token, then only this property is required : enableIdTokenExpiredValidationInRenew. The id_token must be returned in the IDP response otherwise use something like above.

Greetings Damien

**Comment by silviu-sterian** [comment]
Thanks @damienbod! When is 14.2 being released?

Also, my issue had to do with token being renewed (silence renewal triggered) a certain number of seconds before the token expires - using renewTimeBeforeTokenExpiresInSeconds
 I did not see it in your configuration but this property is what was not working for me as doccumented.
 

> This should be fixed., A lot of these issues is due to us validating the id_token and the id_token does not get refreshed and so expires.
> 
> This is fixed in version 14.2.0
> 
> Once released you can configure it something like this:
> 
> ```ts
> import { NgModule } from '@angular/core';
> import { AuthModule, LogLevel } from 'angular-auth-oidc-client';
> 
> @NgModule({
>   imports: [
>     AuthModule.forRoot({
>       config: {
>         authority: '--idp--',
>         redirectUrl: window.location.origin,
>         postLogoutRedirectUri: window.location.origin,
>         clientId: '--client_id--',
>         scope: 'openid profile offline_access',
>         responseType: 'code',
>         silentRenew: true,
>         useRefreshToken: true,
>         ignoreNonceAfterRefresh: true, // this is required if the id_token is not returned
>         allowUnsafeReuseRefreshToken: true, // this is required if the refresh token is not rotated
>         enableIdTokenExpiredValidationInRenew: false, // required to refresh the browser if id_token is not updated after the first authentication
>         autoUserInfo: false, // if the user endpoint is not supported
>         logLevel: LogLevel.Debug,
>       },
>     }),
>   ],
>   exports: [AuthModule],
> })
> export class AuthConfigModule {}
> ```
> 
> If you only need to disable thr id_token check and use only the access token, then only this property is required : enableIdTokenExpiredValidationInRenew. The id_token must be returned in the IDP response otherwise use something like above.
> 
> Greetings Damien
